File: cracking_the_coding_interview_book/chapter_2/partition.cpp
#include <iostream>
#include "LinkedList.cpp"

using namespace std;

void partition(LinkedList* list, int k) {
    Node* head = list->root;
    Node* current = head;
    while (current != NULL && current->next != NULL) {
        if (current->next->data < k) {
            Node* aux = current->next;
            current->next = aux->next;
            aux->next = head;
            head = aux;
        } else {
            current = current->next;
        }
    }
    list->root = head;
}

void test(int* A, int n, int k) {
    LinkedList* list = new LinkedList();
    for (int i = 0; i < n; i++) {
        list->insert(A[i]);
    }
    list->print();
    cout << "Partition: " << k << endl;
    partition(list, k);
    list->print();
    cout << endl;
}

int main() {
    int A[] = {1, 2, 3, 2, 1};
    test(A, 5, 2);
    int B[] = {};
    test(B, 0, 0);
    int C[] = {1, 4, 2, 1};
    test(C, 4, 3);
    int D[] = {1, 2, 3, 5};
    test(D, 4, 5);
    int E[]  = {1, 2, 3, 2, 1, 7};
    test(E, 6, 8);
    return 0;
}

File: legacy_code/linked_list/partition.py
#!/bin/python
from linked_list import create_list, print_list, print_separator, N

def update_list(head, tail, node):
    if not head:
        head = node
        tail = node
    else:
        tail.next = node
        tail = tail.next
    return (head, tail)

def partition(root, x):
    head = head_tail = tail = tail_tail = None
    node = root
    while node:
        if node.value < x:
            head, head_tail = update_list(head, head_tail, node)
        else:
            tail, tail_tail = update_list(tail, tail_tail, node)
        node = node.next
    # Remove tail cycle
    if tail: tail_tail.next = None

    # Merge head and tail
    if head: head_tail.next = tail
    else: head = tail

    # Exist
    return head

print "PARTITION"
x = 50
root = create_list(N)
print print_list(root)
root = partition(root, x)
print print_list(root)
print_separator()
